### PR TITLE
Adjust use of CHPL_BIN_SUBDIR in the skipif and execenvs for this test dir

### PR DESCRIPTION
**Pull request by lydia-duncan**
CHPL_BIN_SUBDIR might not be set in the testing environment, but
CHPL_HOST_BIN_SUBDIR should be available for skipifs.

While here, also use os.path.join instead of concatenating the path
with "/" myself.

We couldn't use the same strategy for the EXECENV due to not injecting
environment variables when pulling in non-executable files.  Elliot has a
PR to fix that, but in the meanwhile mimic the behavior of the mason
testing EXECENVs and explicitly call the chplenv script

Suggested by Elliot

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>

### CODE REVIEW

**Comment by ronawho** [comment]
Does `test/library/packages/ProtobufProtocolSupport/EXECENV` need to be updated too?

File: test/library/packages/ProtobufProtocolSupport/anyRunner.skipif [hunk 1]
```diff
@@ -4,7 +4,7 @@ from distutils.spawn import find_executable
 
 import os
 
-chplPath = os.environ['CHPL_HOME'] + "/bin/" + os.environ['CHPL_BIN_SUBDIR']
+chplPath = os.environ['CHPL_HOME'] + "/bin/" + os.environ['CHPL_HOST_BIN_SUBDIR']
```
Comment (ronawho) [inline test/library/packages/ProtobufProtocolSupport/anyRunner.skipif:7]: Nit, but if we're making changes would be nicer to use `os.path.join`

Comment (lydia-duncan) [reply test/library/packages/ProtobufProtocolSupport/anyRunner.skipif:7]: lol, I GUESS :P

Comment (lydia-duncan) [reply test/library/packages/ProtobufProtocolSupport/anyRunner.skipif:7]: I'm so glad all the other ones symlinks

**Comment by lydia-duncan** [comment]
> Does `test/library/packages/ProtobufProtocolSupport/EXECENV` need to be updated too?

Probably.  It's harder to tell since CHPL_BIN_SUBDIR worked in my environment without me explicitly setting it

**Comment by ronawho** [comment]
> Probably. It's harder to tell since CHPL_BIN_SUBDIR worked in my environment without me explicitly setting it

Do you know what's setting `CHPL_BIN_SUBDIR` in your env? I don't think chplenv knows that var, but it's something I set in my bashrc.

**Comment by lydia-duncan** [comment]
It looks like it's in my .bashrc as well, but in the section I definitely copied from setchplenv.bash

**Comment by lydia-duncan** [comment]
Okay, setting it in the EXECENV doesn't work, the executable environment doesn't seem to know about that variable (at least unless I set it myself)

**Comment by ronawho** [comment]
execenv/skipif and others get the output of `printchplenv --all --simple --no-tidy --internal` injected into their env

File: test/library/packages/ProtobufProtocolSupport/anyRunner.skipif [hunk 2]
```diff
@@ -3,8 +3,9 @@
 from distutils.spawn import find_executable
 
 import os
+import os.path
```
Comment (ronawho) [inline test/library/packages/ProtobufProtocolSupport/anyRunner.skipif:6]: Even more nit so I'm fine leaving it, but I don't think this import is needed

**Comment by lydia-duncan** [comment]
> execenv/skipif and others get the output of `printchplenv --all --simple --no-tidy --internal` injected into their env

Where does that happen?  It seems bad if that only happens for nightly and not start_test itself

**Comment by ronawho** [comment]
> Where does that happen? It seems bad if that only happens for nightly and not start_test itself

That's built into sub_test, nothing to do with nightly -- https://github.com/chapel-lang/chapel/blob/master/util/test/sub_test.py#L268 

Although, looking at it now it looks like we inject the env if the file is executable, but if it's not we just read the file and expand vars with `os.path.expandvars`, so `CHPL_*_BIN_SUBDIR` must not be set in the sub_test env

**Comment by lydia-duncan** [comment]
> > Where does that happen? It seems bad if that only happens for nightly and not start_test itself
> 
> That's built into sub_test, nothing to do with nightly -- https://github.com/chapel-lang/chapel/blob/master/util/test/sub_test.py#L268
> 
> Although, looking at it now it looks like we inject the env if the file is executable, but if it's not we just read the file and expand vars with `os.path.expandvars`, so `CHPL_*_BIN_SUBDIR` must not be set in the sub_test env

Woof.  So do you think I have to make the EXECENV an executable script instead so that it will always get CHPL_HOST_BIN_SUBDIR injected and can use it?

**Comment by ronawho** [comment]
> Woof. So do you think I have to make the EXECENV an executable script instead so that it will always get CHPL_HOST_BIN_SUBDIR injected and can use it?

Yeah, that would be a way to achieve it, but it seems wrong to me that we don't expand chplenv vars when reading non-executable files. I think the following patch will fix that:

<details>

```diff
diff --git a/util/test/sub_test.py b/util/test/sub_test.py
index 071eabffb4..99a5b14c9d 100755
--- a/util/test/sub_test.py
+++ b/util/test/sub_test.py
@@ -261,6 +261,21 @@ def PerfDirFile(s):
 def PerfTFile(test_filename, sfx):
     return test_filename + '.' + perflabel + sfx

+def get_chplenv():
+    env_cmd = [os.path.join(utildir, 'printchplenv'), '--all', '--simple', '--no-tidy', '--internal']
+    chpl_env = py3_compat.Popen(env_cmd, stdout=subprocess.PIPE).communicate()[0]
+    chpl_env = dict(map(lambda l: l.split('=', 1), chpl_env.splitlines()))
+    return chpl_env
+
+# os.path.expandvars but stuff chplenv into it too
+def expandvars_chpl(path):
+    orig_env = dict(os.environ)
+    os.environ.update(get_chplenv())
+    new_path = os.path.expandvars(path)
+    os.environ.clear()
+    os.environ.update(orig_env)
+    return new_path
+
 # Read a file or if the file is executable read its output. If the file is
 # executable, the current chplenv is copied into the env before executing.
 # Expands shell variables and strip out comments/whitespace. Returns a list of
@@ -272,9 +287,7 @@ def ReadFileWithComments(f, ignoreLeadingSpace=True):
     if os.access(f, os.X_OK):
         try:
             # grab the chplenv so it can be stuffed into the subprocess env
-            env_cmd = [os.path.join(utildir, 'printchplenv'), '--all', '--simple', '--no-tidy', '--internal']
-            chpl_env = py3_compat.Popen(env_cmd, stdout=subprocess.PIPE).communicate()[0]
-            chpl_env = dict(map(lambda l: l.split('=', 1), chpl_env.splitlines()))
+            chpl_env = get_chplenv()
             file_env = os.environ.copy()
             file_env.update(chpl_env)

@@ -303,7 +316,7 @@ def ReadFileWithComments(f, ignoreLeadingSpace=True):
         else:
             if line[0] == '#': continue
         # expand shell variables
-        line = os.path.expandvars(line)
+        line = expandvars_chpl(line)
         mylist.append(line)
     return mylist

```

</details>

Looking a little more, skipifs are processed differently with `runSkipIf`, which uses `util/test/testEnv` which injects the chplenv for executable and non-executable so to me it feels like we should do that for non-executable files read with `ReadFileWithComments` too

I don't think I'll have time to run with that patch this week, but if you do feel free to use it. I tested it with a fake example using `CHPL_HOST_BIN_PLATFORM`, but didn't do any significant testing on it. I feel semi-confident in it, though the comment above it probably needs updating.

**Comment by lydia-duncan** [comment]
> Yeah, that would be a way to achieve it, but it seems wrong to me that we don't expand chplenv vars when reading non-executable files.

Yeah, I agree.  I'm more coming at it from the perspective of "this PR was from last sprint cycle and there's plenty of other things I need to do this sprint cycle", so would prefer to make an issue rather than being the one to fix it at the cost of my other tasks.

**Comment by ronawho** [comment]
>> Yeah, that would be a way to achieve it, but it seems wrong to me that we don't expand chplenv vars when reading non-executable files.

> Yeah, I agree. I'm more coming at it from the perspective of "this PR was from last sprint cycle and there's plenty of other things I need to do this sprint cycle", so would prefer to make an issue rather than being the one to fix it at the cost of my other tasks.

Ok, no problem -- I opened https://github.com/chapel-lang/chapel/pull/16812 for that patch (I tend to find that it takes me longer to create an issue others can understand and ramp back up on the problem when I finally have to come back to it compared to just doing it.)

**Review by ronawho — APPROVED** [review]

**Pull request merged**